More FV PorousFlow

cpgr opened the pull request:
This adds some more FV functionality to the PorousFlow module.

Specifically, there is a lot more boilerplate templating of materials to allow them to work with AD variables as well as non-AD variables, another FV kernel is added (the FV equivalent to PorousFlowDispersiveFlux, and more FV tests (FV equivalents of existing FE tests). 

Refs #21275

cpgr committed:
Template PorousFlowDarcyVelocityComponent for AD

This allows this AuxVariable to be used in FV models.

Refs #21275

cpgr committed:
Template materials that provide pressure and saturation

These changes will enable these materials to be used in FV
PorousFlow simulations

Refs #21275

cpgr committed:
Template mass fraction material for AD

This will enable this material to be used in FV PorousFlow
simulations

Refs #21275

cpgr committed:
Add missing AD method to PorousFlowCapillaryPressure

Refs #21275

cpgr committed:
Template diffusivity materials for AD

Refs #21275

cpgr committed:
Template relative permeability materials

Refs #21275

cpgr committed:
Use reconstructed gradient on boundary elements

Refs #21275

cpgr committed:
FV dispersive flux kernel

Refs #21275

cpgr committed:
getfieldvar

cpgr committed:
getVar to getFieldVar to support FE and FV variables

Refs #21275

cpgr committed:
Test finite volume diffusive and dispersive flux

Refs #21275

cpgr committed:
Documentation for FVPorousFlowDispersiveFlux

Refs #21275

cpgr committed:
Clang-format EOL fix

Refs #21275

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1567188/) on a226375 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24582/site)

This comment will be updated on new commits.

cpgr committed:
Increase ghosting to avoid assertion in parallel

It seems that these FVKernels need 2 host layers to run in
parallel.

Refs #21275

cpgr committed:
Add missing template class in PorousFlowDiffusivityBase

Refs #21275

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1567380/) on 84ea4ed wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>968eb4</th>
      <th colspan=4>#24582 84ea4e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.40%</td>
      <td>85.39%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88702</td>
      <td>88699</td>
      <td>-3</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15168</td>
      <td>15171</td>
      <td>+3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24582/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24582/coverage/framework)


## Modules coverage

### Porous flow

<table>
  <thead>
    <tr>
      <th></th>
      <th>968eb4</th>
      <th colspan=4>#24582 84ea4e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>95.91%</td>
      <td>95.60%</td>
      <td>-0.31%</td>
      <td>88.01%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10715</td>
      <td>10780</td>
      <td>+65</td>
      <td>301</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>457</td>
      <td>496</td>
      <td>+39</td>
      <td>41</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24582/coverage_diff/porous_flow)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24582/coverage/porous_flow)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24582/coverage/xfem)

</details>

## Warnings

- `porous_flow` new line coverage rate 88.01% is less than the suggested 90.0%



This comment will be updated on new commits.

cpgr committed:
Unit tests for AD relperm utils

Refs #21275

cpgr committed:
Check that fluid component is valid in kernels

Refs #21275

cpgr committed:
Add FV test equivalents of more pressure pulse tests

These tests test the FVKernels as well as the the AD version
of the materials in PorousFlow.

Refs #21275

cpgr committed:
FV versions of gravity equilibrium tests

Refs #21275

cpgr committed:
Make unit test name consistent

Refs #21275

cpgr commented:
Hmm, I can't figure out this error in the `Newer gcc` tests. 

```
               from /data/civet1/build/moose/modules/porous_flow/unit/src/PorousFlowFLACrelpermTest.C:12:
In member function 'MetaPhysicL::DynamicSparseNumberBase<Data, Indices, SubType, SubTypeArgs>::value_type& MetaPhysicL::DynamicSparseNumberBase<Data, Indices, SubType, SubTypeArgs>::insert(unsigned int) [with Data = MetaPhysicL::DynamicStdArrayWrapper<double, MetaPhysicL::NWrapper<53> >; Indices = MetaPhysicL::DynamicStdArrayWrapper<long unsigned int, MetaPhysicL::NWrapper<53> >; SubType = MetaPhysicL::SemiDynamicSparseNumberArray; SubTypeArgs = {double, long unsigned int, MetaPhysicL::NWrapper<53>}]',
    inlined from 'void Moose::derivInsert(MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<N> >&, libMesh::dof_id_type, libMesh::Real) [with long unsigned int N = 53]' at /data/civet1/build/moose/framework/build/header_symlinks/SparseOps.h:36:16,
    inlined from 'virtual void PorousFlowFLACrelpermTest_adrelperm_Test::TestBody()' at /data/civet1/build/moose/modules/porous_flow/unit/src/PorousFlowFLACrelpermTest.C:57:21:
/opt/libmesh/include/metaphysicl/dynamicsparsenumberbase.h:220:17: error: 'sat' may be used uninitialized [-Werror=maybe-uninitialized]
  220 |       *upper_it != I(i))
      |       ~~~~~~~~~~^~~~~~~
/data/civet1/build/moose/modules/porous_flow/unit/src/PorousFlowFLACrelpermTest.C: In member function 'virtual void PorousFlowFLACrelpermTest_adrelperm_Test::TestBody()':
/data/civet1/build/moose/modules/porous_flow/unit/src/PorousFlowFLACrelpermTest.C:56:12: note: 'sat' declared here
   56 |   DualReal sat = 0.3;
```

The offending code looks ok to me?
```
TEST(PorousFlowFLACrelpermTest, adrelperm)
 {
   DualReal sat = 0.3;
   Moose::derivInsert(sat.derivatives(), 0, 1.0);

   const auto adrelperm = PorousFlowFLACrelperm::relativePermeability(sat, 2.7);

   const auto relperm = PorousFlowFLACrelperm::relativePermeability(sat.value(), 2.7);
   const auto drelperm = PorousFlowFLACrelperm::dRelativePermeability(sat.value(), 2.7);

   EXPECT_NEAR(adrelperm.value(), relperm, 1.0E-5);
   EXPECT_NEAR(adrelperm.derivatives()[0], drelperm, 1.0E-5);
 }
```

cpgr commented on modules/porous_flow/src/fvkernels/FVPorousFlowDispersiveFlux.C:
```diff
@@ -0,0 +1,175 @@
+      diffusion = _porosity[_qp] * _tortuosity[_qp][p] * _density[_qp][p] *
+                  _diffusion_coeff[_qp][p][_fluid_component];
+    }
+    else
```
Hi @lindsayad 

Can I get your thoughts on this bit please? If you remember from the first bit of FV that you helped me with, ghosting was removed so that there was an error when trying to access neighbour properties on a boundary element (face?). To get around it, you made a BC that added an flux BC and a Dirichlet BC (I think the logic is that using a FluxBC disables the evaluation of the kernel on the boundary face). 

However, I realised that I could use the gradient of things that we compute in materials on these edge elements (we are computing them anyway), so added the 
```
if (on boundary)
   use gradient from material
else
   compute gradient using neighbour element
```
logic you see here. This makes the FV input files just like the existing FE inputs (don't need an additional BC), and it avoids having to make a really complex BC that covers all of our cases. It seems to work ok, and I believe that if we do add a FluxBC like the one you made, that it wouldn't evaluate the `on boundary` bit due to how the FV kernels are disabled on boundaries when a FLuxBC is present.

Is my understanding correct here?

lindsayad replied:
Yes I think you're correct (a small clarification question below) and what you're doing seems logical to me. It would definitely be nicer to not have to add the additional object and have the 1-to-1 relationship with FE inputs.

So you have a Dirichlet BC right? That's the only way you will enter `computeQpResidual` for your `FVFluxKernel` object. As you correctly stated, we do not apply kernels on boundaries where FVFluxBCs reside. And we also do not apply them if *no* BCs are applied. E.g. the only time a `FVFluxKernel` runs on a boundary (outside of some forcing parameters) is if a Dirichlet BC is applied.

cpgr replied:
OK, I think that is what I want. Use the reconstructed gradient when there is a Dirichlet BC, don't evaluate the kernel on the boundary when it is a natural boundary (no BC), and also don't evaluate the kernel on the boundary when there is a FluxBC. Thanks for the clarification.

cpgr commented:
It looks like the `Newer gcc` unit test error has defeated me! 

@lindsayad - could you take a look at this PR when you get a chance? Like I mentioned earlier, most of it is just templating our existing materials and adding tests. The only new object is the FVKernel.

lindsayad commented on modules/porous_flow/test/tests/gravity/grav02b_fv.i:
```diff
@@ -0,0 +1,213 @@
+    full = true
+  []
+[]
+
```
```suggestion
```

lindsayad commented on modules/porous_flow/test/tests/gravity/grav02e_fv.i:
```diff
@@ -0,0 +1,226 @@
+    type = SMP
+    full = true
+    petsc_options_iname = '-ksp_type -pc_type -snes_atol -snes_rtol'
+    petsc_options_value = 'bcgs bjacobi 1E-12 1E-10'
```
bcgs, fancy!

lindsayad replied:
how do you choose when to do `bcgs` vs `gmres` ?

cpgr replied:
Haha - copy and paste from an existing test! It probably doesn't even need any if these options to work - I'll see if I can remover them

cpgr replied:
I think it is ok to just delete these options

lindsayad commented on modules/porous_flow/test/tests/pressure_pulse/pressure_pulse_1d_2phase_fv.i:
```diff
@@ -0,0 +1,245 @@
+    full = true
+    petsc_options = '-snes_converged_reason -ksp_diagonal_scale -ksp_diagonal_scale_fix -ksp_gmres_modifiedgramschmidt -snes_linesearch_monitor'
+    petsc_options_iname = '-ksp_type -pc_type -sub_pc_type -sub_pc_factor_shift_type -pc_asm_overlap -snes_atol -snes_rtol -snes_max_it'
+    petsc_options_value = 'gmres      asm      lu           NONZERO                   2               1E-15       1E-20 20'
```
are these tight tolerances necessary?

cpgr replied:
Also copy and paste. I'll see if they are really needed

cpgr replied:
Looks like these aren't needed

lindsayad reviewed (changes requested):
really impressive work

cpgr commented on modules/porous_flow/test/tests/dispersion/diff01_fv.i:
```diff
@@ -0,0 +1,200 @@
+
+[Variables]
+  [pp]
+    family = MONOMIAL
+    order = CONSTANT
+    fv = true
```
Oh, I didn't know this!

cpgr commented on modules/porous_flow/src/materials/PorousFlow1PhaseP.C:
```diff
@@ -64,30 +63,28 @@ PorousFlow1PhasePTempl<is_ad>::computeQpProperties()
-            _gradp_qp_var[_qp];
-      }
+      (*_dgradp_qp_dgradv)[_qp][0][_p_var_num] = 1.0;
+      (*_dgrads_qp_dgradv)[_qp][0][_p_var_num] = MetaPhysicL::raw_value(ds);
+      (*_dgrads_qp_dv)[_qp][0][_p_var_num] =
+          _pc_uo.d2Saturation(MetaPhysicL::raw_value(_porepressure_var[_qp])) *
+          MetaPhysicL::raw_value(_gradp_qp_var[_qp]);
```
I was getting compile time errors without the `raw_value`'s, but looking into `constexp` it looks like it might avoid these. I'll give it a try. Thanks!

cpgr replied:
Yep, this worked!

cpgr commented on modules/porous_flow/include/userobjects/PorousFlowCapillaryPressure.h:
```diff
@@ -60,6 +60,7 @@ class PorousFlowCapillaryPressure : public DiscreteElementUserObject
    * @return derivative of capillary pressure with respect to true saturation
    */
   virtual Real dCapillaryPressure(Real saturation, unsigned qp = 0) const;
+  virtual DualReal dCapillaryPressure(const DualReal & saturation, unsigned qp = 0) const;
```
I probably use these interchangeably! I'll change them to `ADReal`

cpgr committed:
Template relative permeability methods

Reduce duplication by templating relativePermeability().

Refs #21275

cpgr committed:
DualReal to ADReal

Use ADReal for consistency

Refs #21275

cpgr committed:
Don't copy AD variables

Refs #21275

cpgr committed:
Use constexp to reduce use of raw_value

Refs #21275

cpgr committed:
Clean up new tests

Refs #21275

cpgr commented on modules/porous_flow/src/userobjects/PorousFlowCapillaryPressure.C:
```diff
@@ -68,6 +68,18 @@ PorousFlowCapillaryPressure::capillaryPressure(Real saturation, unsigned qp) con
+  const Real dPc_ds = dCapillaryPressure(saturation.value(), qp);
+
+  DualReal result = Pc;
+  result.derivatives() = saturation.derivatives() * dPc_ds;
```
I'll look into refactoring this later to see if I can avoid duplication

cpgr commented on modules/porous_flow/test/tests/gravity/grav01a_fv.i:
```diff
@@ -0,0 +1,198 @@
+  []
+[]
+
+[Preconditioning]
+  [smp]
+    type = SMP
+    full = true
+  []
+[]
```
I'll leave it in but good to know

cpgr commented on modules/porous_flow/src/utils/PorousFlowBrooksCorey.C:
```diff
@@ -90,6 +90,17 @@ relativePermeabilityW(Real seff, Real lambda)
+    return 1.0;
+
+  return std::pow(seff, (2.0 + 3.0 * lambda) / lambda);
+}
```
I templated these. They had to go in the header though, not the .C file, or the compiler complained about missing symbols

cpgr committed:
Need libmesh/utility.h in header

Utility::pow() needs to be available in the header now.

Refs #21275

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1579090/) on 0418302 : invalidated by @lindsayad

lindsayad committed:
Workaround 'possibly unitialized' warning

We don't understanding this warning as we have this idiom in numerous
other places in our code

moosebuild commented:
Job [Documentation internal apps](https://civet.inl.gov/job/1580419/) on 0c45e20 : invalidated by @lindsayad

cpgr commented:
Thanks so much @lindsayad

lindsayad commented:
You're welcome and thank you for the always great contributions!
